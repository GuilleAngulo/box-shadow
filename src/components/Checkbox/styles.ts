import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
  margin: 0.6rem 0;
`

export const Input = styled.input`
  ${({ theme }) => css`
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: center;
    appearance: none;
    width: 1.8rem;
    height: 1.8rem;
    border: 0.2rem solid ${theme.colors.font};
    border-radius: 0.2rem;
    transition: background border ${theme.transition.fast};
    position: relative;
    outline: none;

    &:before {
      content: '';
      width: 0.6rem;
      height: 0.9rem;
      border: 0.2rem solid white;
      border-top: 0;
      border-left: 0;
      transform: rotate(45deg);
      position: absolute;
      top: 0.1rem;
      opacity: 0;
      transition: ${theme.transition.fast};
    }

    &:focus {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    &:hover {
      border-color: ${theme.colors.border};
      transition: ${theme.transition.fast};
    }

    &:checked {
      border-color: ${theme.colors.primary};
      background: ${theme.colors.primary};

      &:before {
        opacity: 1;
      }
    }
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    cursor: pointer;
    padding-left: ${theme.spacings.xxsmall};
    line-height: 1.8rem;
    color: ${theme.colors.font};
  `}
`
