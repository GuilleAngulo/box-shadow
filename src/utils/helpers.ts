import parserPostcss from 'prettier/parser-postcss'
import prettier from 'prettier/standalone'
import { ShadowProps } from 'types'

export const prettify = (code: string) => {
  const css = code && `box-shadow: ${code}`
  return prettier.format(css, {
    parser: 'css',
    printWidth: 70,
    plugins: [parserPostcss]
  })
}

export const stringify = (shadowStyle: ShadowProps[], semicolon = true) => {
  const stylesArray =
    Array.isArray(shadowStyle) &&
    shadowStyle.map(
      ({
        horizontalOffset,
        verticalOffset,
        blurRadius,
        spreadRadius,
        color,
        inset = false
      }) =>
        `${
          inset ? 'inset ' : ''
        }${horizontalOffset}px ${verticalOffset}px ${blurRadius}px ${spreadRadius}px rgba(${
          color.red
        }, ${color.green}, ${color.blue}, ${color.alpha})`
    )

  let styles = Array.isArray(stylesArray) ? stylesArray.join(', ') : ''

  if (semicolon) {
    styles += ';'
  }

  return styles
}
