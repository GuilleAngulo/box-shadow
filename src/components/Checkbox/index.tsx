import { InputHTMLAttributes } from 'react'
import * as S from './styles'

export type CheckboxProps = {
  onCheck?: (status: boolean) => void
  isChecked?: boolean
  label?: string
  labelFor?: string
  value?: string | ReadonlyArray<string> | number
} & InputHTMLAttributes<HTMLInputElement>

const Checkbox = ({
  onCheck,
  isChecked = false,
  label,
  labelFor = '',
  value,
  ...props
}: CheckboxProps) => {
  const onChange = () => {
    const status = !isChecked

    if (onCheck) {
      onCheck(status)
    }
  }

  return (
    <S.Wrapper>
      <S.Input
        id={labelFor}
        type="checkbox"
        onChange={onChange}
        checked={isChecked}
        value={value}
        {...props}
      />
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
    </S.Wrapper>
  )
}

export default Checkbox
