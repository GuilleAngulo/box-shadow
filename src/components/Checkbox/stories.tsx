import { Story, Meta } from '@storybook/react/types-6-0'
import { useState } from 'react'
import Checkbox, { CheckboxProps } from '.'

export default {
  title: 'Form/Checkbox',
  component: Checkbox,
  parameters: {
    layout: 'fullscreen'
  },
  argTypes: {
    onCheck: { action: 'checked' }
  }
} as Meta

export const Default: Story<CheckboxProps> = (args) => {
  const [value, setValue] = useState(false)
  return (
    <div style={{ padding: 10 }}>
      <Checkbox
        name="category"
        label="Action"
        labelFor="action"
        isChecked={value}
        onCheck={setValue}
        {...args}
      />
    </div>
  )
}
