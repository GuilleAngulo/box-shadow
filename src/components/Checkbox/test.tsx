import { screen, waitFor } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import userEvent from '@testing-library/user-event'

import Checkbox from '.'

describe('<Checkbox />', () => {
  it('should render with label', () => {
    const { container } = renderWithTheme(
      <Checkbox label="checkbox label" labelFor="check" />
    )

    expect(screen.getByRole('checkbox')).toBeInTheDocument()
    expect(screen.getByLabelText(/checkbox label/i)).toBeInTheDocument()
    expect(screen.getByText(/checkbox label/i)).toHaveAttribute('for', 'check')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('it should render without label', () => {
    renderWithTheme(<Checkbox />)

    expect(screen.queryByLabelText('Checkbox')).not.toBeInTheDocument()
  })

  it('it should render with black label', () => {
    renderWithTheme(
      <Checkbox label="checkbox label" labelFor="check" labelColor="black" />
    )

    expect(screen.getByText(/checkbox label/i)).toHaveStyle({
      color: '#030517'
    })
  })

  it('should dispatch onCheck when status changes', async () => {
    const onCheck = jest.fn()

    renderWithTheme(<Checkbox label="Checkbox" onCheck={onCheck} />)

    expect(onCheck).not.toHaveBeenCalled()

    userEvent.click(screen.getByRole('checkbox'))
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1)
    })
    expect(onCheck).toHaveBeenCalledWith(true)
  })

  it('should dispatch onCheck when status changes', async () => {
    const onCheck = jest.fn()

    renderWithTheme(<Checkbox label="Checkbox" onCheck={onCheck} isChecked />)

    expect(onCheck).not.toHaveBeenCalled()

    userEvent.click(screen.getByRole('checkbox'))
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1)
    })
    expect(onCheck).toHaveBeenCalledWith(false)
  })

  it('should be accesible with tab', () => {
    renderWithTheme(<Checkbox label="Checkbox" labelFor="Checkbox" />)

    expect(document.body).toHaveFocus()

    userEvent.tab()

    expect(screen.getByLabelText(/checkbox/i)).toHaveFocus()
  })
})
