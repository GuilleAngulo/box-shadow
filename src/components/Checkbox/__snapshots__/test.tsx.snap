// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render with label 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 0.8rem;
}

.c1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 1.8rem;
  height: 1.8rem;
  border: 0.2rem solid #818181;
  border-radius: 0.2rem;
  -webkit-transition: all var(--transition-fast);
  transition: all var(--transition-fast);
  -webkit-transition-property: background,border;
  transition-property: background,border;
  position: relative;
  outline: none;
}

.c1:before {
  content: '';
  width: 0.6rem;
  height: 0.9rem;
  border: 0.2rem solid white;
  border-top: 0;
  border-left: 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  position: absolute;
  top: 0.1rem;
  opacity: 0;
  -webkit-transition: var(--transition-fast);
  transition: var(--transition-fast);
}

.c1:focus {
  box-shadow: 0 0 0.5rem var(--color-primary);
}

.c1:hover {
  border-color: #7D7D7D;
  -webkit-transition: var(--transition-fast);
  transition: var(--transition-fast);
}

.c1:checked {
  border-color: var(--color-primary);
  background: var(--color-primary);
}

.c1:checked:before {
  opacity: 1;
}

.c2 {
  cursor: pointer;
  padding-left: var(--spacings-xxsmall);
  line-height: 1.8rem;
  color: #383838;
  font-weight: var(--font-bold);
}

<div
  class="c0"
>
  <input
    class="c1"
    id="check"
    type="checkbox"
    value=""
  />
  <label
    class="c2"
    for="check"
  >
    checkbox label
  </label>
</div>
`
